sklearn/tests/test_base.py

```diff
@@ -562,19 +565,20 @@ def test_pickle_version_warning_is_issued_when_no_version_info_in_pickle():
 
 
 def test_pickle_version_no_warning_is_issued_with_non_sklearn_estimator():
+    # Avoid side-effect on TreeNoVersion class
+    tree_class = copy.copy(TreeNoVersion)
+    tree_class.__module__ = "notsklearn"
+    # Need this for the class to be picklable
+    sys.modules["notsklearn"] = types.ModuleType("notsklearn")
+    sys.modules["notsklearn"].TreeNoVersion = tree_class
+
     iris = datasets.load_iris()
-    tree = TreeNoVersion().fit(iris.data, iris.target)
+    tree = tree_class().fit(iris.data, iris.target)
     tree_pickle_noversion = pickle.dumps(tree)
-    try:
-        module_backup = TreeNoVersion.__module__
-        TreeNoVersion.__module__ = "notsklearn"
-
-        with warnings.catch_warnings():
-            warnings.simplefilter("error")
 
-            pickle.loads(tree_pickle_noversion)
-    finally:
-        TreeNoVersion.__module__ = module_backup
+    with warnings.catch_warnings():
+        warnings.simplefilter("error")
+        pickle.loads(tree_pickle_noversion)
```

Comment:
We should probably wrap the body of this test in a try/finally construct to remove `types.ModuleType("notsklearn")` from `sys.modules` upon completion/error and furthermore mark this test as thread unsafe (because of the global `sys.modules` side effect).

Reply:
I don't know why exactly (maybe just for the challenge :sweat_smile:) but I was trying to avoid marking the test as thread-unsafe. If we mark the test as thread-unsafe, we may as well keep the previous way of doing it that modifies `TreeNoVersion.__module__` directly?

Reply:
As you prefer. No strong opinion.